src/lib.rs

```diff
@@ -303,3 +191,167 @@ fn generate_analysis(config: &Config) -> Result<(), Box<std::error::Error>> {
 
     Ok(())
 }
+
+#[derive(Debug)]
+struct DocData {
+    krate: Crate,
+    data: HashMap<String, Item>,
+}
+
+impl DocData {
+    fn new(config: &Config) -> Result<DocData, Box<std::error::Error>> {
+        let roots = config.host.def_roots()?;
+
+        // FIXME: this whole code shouldn't look for a precise crate.
+        let root_id = roots.iter().find(|&&(_, ref name)| name == "example");
+        let root_id = match root_id {
+            Some(&(id, _)) => id,
+            _ => return Err(Box::new(CrateErr::new("example"))),
+        };
+
+        let root_def = config.host.get_def(root_id)?;
+
+        let name_len = root_def.qualname.len();
+        let mut krate = Crate {
+            id: root_id,
+            // example:: -> example
+            name: root_def.qualname[..(name_len - 2)].to_string(),
+            docs: root_def.docs.clone(),
+            modules: Vec::new(),
+        };
+
+        let data = DocData::build_data(config, root_id, &mut krate)?;
+
+        Ok(DocData { krate, data })
+    }
+
+    fn build_data(
+        config: &Config,
+        root_id: analysis::Id,
+        krate: &mut Crate,
```

Comment:
this is slightly awkward, but is needed since we actually figure out what goes in the crate once we try to process all of the data.

we could instead return everything, I guess... dunno